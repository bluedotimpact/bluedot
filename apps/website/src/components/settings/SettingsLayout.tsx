import React from 'react';
import { ClickTarget, NewText, BluedotRoute } from '@bluedot/ui';
import clsx from 'clsx';
import { H2 } from '../Text';

type SettingsTab = 'account' | 'courses' | 'community';

type SettingsLayoutProps = {
  activeTab: SettingsTab;
  children: React.ReactNode;
  route: BluedotRoute;
};

const SettingsLayout = ({ activeTab, children, route }: SettingsLayoutProps) => {
  return (
    <>
      {/* Breadcrumbs */}
      <SettingsBreadcrumbs route={route} />

      <div className="container mx-auto px-4 py-6">
        {/* Mobile and Tablet layout - horizontal navigation */}
        <div className="lg:grid lg:grid-cols-12 lg:gap-8">
          {/* Navigation - horizontal on mobile/tablet, sidebar on desktop */}
          <aside className="lg:col-span-3 mb-6 lg:max-w-[320px]" aria-label="Settings sidebar">
            <H2 className="mb-4 font-semibold text-size-lg leading-[1.4]">Settings</H2>
            <SettingsNavigation activeTab={activeTab} />
          </aside>

          {/* Content area */}
          <main className="lg:col-span-9" aria-label="Settings content">
            <div className="container-lined" role="region">
              {children}
            </div>
          </main>
        </div>
      </div>
    </>
  );
};

const SettingsNavigation = ({ activeTab }: { activeTab: SettingsTab }) => {
  const navigationItems = [
    {
      id: 'account' as SettingsTab,
      label: 'Account',
      href: '/settings/account',
    },
    {
      id: 'courses' as SettingsTab,
      label: 'Courses',
      href: '/settings/courses',
    },
    {
      id: 'community' as SettingsTab,
      label: 'Community',
      href: '/settings/community',
      isNew: true,
    },
  ];

  return (
    <nav aria-label="Settings navigation">
      <ul className="flex gap-2 flex-wrap lg:block lg:space-y-1">
        {navigationItems.map((item) => {
          const isActive = activeTab === item.id;
          return (
            <li key={item.id}>
              <ClickTarget
                url={item.href}
                className={`
                  w-auto lg:w-full h-11 px-4 py-3 rounded-lg transition-colors flex items-center
                  ${isActive
                  ? 'bg-[rgba(0,85,255,0.08)] text-bluedot-darker font-semibold'
                  : 'text-bluedot-darker hover:bg-[rgba(0,85,255,0.04)]'
                  }
                `}
                aria-current={isActive ? 'page' : undefined}
                aria-label={`${item.label} settings${item.isNew ? ' (New feature)' : ''}`}
              >
                <span className="flex items-center gap-3 w-full">
                  <NewText.P className={`text-[13px] leading-[22px] flex-grow !m-0 ${isActive ? 'font-semibold' : 'font-normal'}`}>
                    {item.label}
                  </NewText.P>
                  {item.isNew && (
                    <span
                      className="text-[10px] px-1.5 py-0.5 rounded-[5px] uppercase font-bold bg-[rgba(0,85,255,0.08)] text-[#0037FF]"
                      aria-label="New feature"
                    >
                      NEW
                    </span>
                  )}
                </span>
              </ClickTarget>
            </li>
          );
        })}
      </ul>
    </nav>
  );
};

const SettingsBreadcrumbs = ({ route }: { route: BluedotRoute }) => {
  const breadcrumbItems = [...(route.parentPages ?? []), route];

  return (
    <div className="breadcrumbs bg-color-canvas border-b border-color-divider w-full py-space-between">
      <nav
        className="breadcrumbs__nav section-base flex flex-row justify-between"
        aria-label="Breadcrumbs"
      >
        <ol className="breadcrumbs__list flex">
          {breadcrumbItems.map((item, index) => {
            const isLast = index === breadcrumbItems.length - 1;
            const isSettingsItem = item.url === '/settings';

            return (
              <li key={item.url} className="breadcrumbs__item flex items-center">
                {isSettingsItem ? (
                  <NewText.P
                    className={clsx(
                      'breadcrumbs__link !m-0',
                      isLast && 'opacity-50',
                    )}
                    aria-current={isLast ? 'page' : undefined}
                  >
                    {item.title}
                  </NewText.P>
                ) : (
                  <NewText.A
                    className={clsx(
                      'breadcrumbs__link no-underline',
                      isLast && 'opacity-50',
                    )}
                    href={item.url}
                    aria-current={isLast ? 'page' : undefined}
                  >
                    {item.title}
                  </NewText.A>
                )}
                {!isLast && (
                  <span className="breadcrumbs__separator mx-2" aria-hidden="true">{'>'}</span>
                )}
              </li>
            );
          })}
        </ol>
      </nav>
    </div>
  );
};

export default SettingsLayout;
