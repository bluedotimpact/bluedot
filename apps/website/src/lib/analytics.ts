export const GA_MEASUREMENT_ID = process.env.NEXT_PUBLIC_GA_MEASUREMENT_ID;
export const GA_DEBUG = process.env.NEXT_PUBLIC_GA_DEBUG;

export const isDebugModeEnabled = () => {
  return !!GA_DEBUG;
};

export const isAnalyticsEnabled = () => {
  const hasMeasurementId = !!GA_MEASUREMENT_ID;

  return hasMeasurementId;
};
