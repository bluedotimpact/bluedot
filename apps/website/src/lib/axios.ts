import axios from 'axios';
import { configure } from 'axios-hooks';

// Create axios instance with base URL
const axiosInstance = axios.create({
  baseURL: typeof window !== 'undefined' ? window.location.origin : 'http://localhost:8000',
});

// Configure axios-hooks to use our configured instance
configure({ axios: axiosInstance });

export default axiosInstance;
