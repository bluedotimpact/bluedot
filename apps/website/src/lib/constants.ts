export const FOAI_COURSE_ID = 'rec0Zgize0c4liMl5';
