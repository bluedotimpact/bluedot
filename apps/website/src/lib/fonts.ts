import localFont from 'next/font/local';

// Inter for body text (local files)
export const inter = localFont({
  src: [
    {
      path: '../../public/fonts/Inter-Regular.woff2',
      weight: '400',
      style: 'normal',
    },
    {
      path: '../../public/fonts/Inter-Medium.woff2',
      weight: '500',
      style: 'normal',
    },
    {
      path: '../../public/fonts/Inter-SemiBold.woff2',
      weight: '600',
      style: 'normal',
    },
    {
      path: '../../public/fonts/Inter-Bold.woff2',
      weight: '700',
      style: 'normal',
    },
  ],
  variable: '--font-inter',
  display: 'swap',
});

// Inter Display for headlines (local files)
export const interDisplay = localFont({
  src: [
    {
      path: '../../public/fonts/InterDisplay-Regular.woff2',
      weight: '400',
      style: 'normal',
    },
    {
      path: '../../public/fonts/InterDisplay-Medium.woff2',
      weight: '500',
      style: 'normal',
    },
    {
      path: '../../public/fonts/InterDisplay-SemiBold.woff2',
      weight: '600',
      style: 'normal',
    },
    {
      path: '../../public/fonts/InterDisplay-Bold.woff2',
      weight: '700',
      style: 'normal',
    },
  ],
  variable: '--font-inter-display',
  display: 'swap',
});
