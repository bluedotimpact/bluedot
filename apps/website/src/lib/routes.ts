import type { BluedotRoute } from '@bluedot/ui';

const login: BluedotRoute = {
  title: 'Login',
  url: '/login',
};

const home: BluedotRoute = {
  title: 'Home',
  url: '/',
};

const about: BluedotRoute = {
  title: 'About us',
  url: '/about',
  parentPages: [home],
};

const privacyPolicy: BluedotRoute = {
  title: 'Privacy Policy',
  url: '/privacy-policy',
  parentPages: [home],
};

const joinUs: BluedotRoute = {
  title: 'Join us',
  url: '/join-us',
  parentPages: [home],
};

// Avoid changing the /contact route - we reference it lots of places.
// If you are changing it, update libraries/ui/src/constants.ts
const contact: BluedotRoute = {
  title: 'Contact us',
  url: '/contact',
  parentPages: [home],
};

const certification: BluedotRoute = {
  title: 'Certification',
  url: '/certification',
  parentPages: [home],
};

const courses: BluedotRoute = {
  title: 'Courses',
  url: '/courses',
  parentPages: [home],
};

const profile: BluedotRoute = {
  title: 'Profile',
  url: '/profile',
  parentPages: [home],
};

const logout: BluedotRoute = {
  title: 'Log out',
  url: '/login/clear',
  parentPages: [home],
};

export const ROUTES = {
  about,
  certification,
  contact,
  courses,
  home,
  joinUs,
  login,
  logout,
  privacyPolicy,
  profile,
} as const;
