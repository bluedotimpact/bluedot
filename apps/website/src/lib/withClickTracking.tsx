import React from 'react';
import { sendGAEvent } from '@next/third-parties/google';

// Base tracking configuration type
type TrackingConfig = {
  eventName: string;
  eventParams?: Record<string, string>;
};

// Props that will be available on the wrapped component
type TrackingProps = {
  trackingEventName?: string;
  trackingEventParams?: Record<string, string>;
};

// HOC function
export const withClickTracking = <P extends object>(
  WrappedComponent: React.ComponentType<P>,
  defaultConfig: TrackingConfig = { eventName: 'click', eventParams: {} },
) => {
  // Return wrapped component that accepts both original props and tracking props
  const ClickTrackingComponent = (
    props: P & TrackingProps,
  ) => {
    const {
      trackingEventName,
      trackingEventParams,
      ...componentProps
    } = props;

    // Merge default config with props passed to the component
    const eventConfig = {
      eventName: trackingEventName || defaultConfig.eventName,
      eventParams: {
        ...defaultConfig.eventParams,
        ...trackingEventParams,
      },
    };

    const handleClick = () => {
      sendGAEvent('event', eventConfig.eventName, eventConfig.eventParams);
    };

    // eslint-disable-next-line jsx-a11y/click-events-have-key-events, jsx-a11y/no-static-element-interactions
    return <span onClick={handleClick}><WrappedComponent {...componentProps as P} /></span>;
  };
  return ClickTrackingComponent;
};
