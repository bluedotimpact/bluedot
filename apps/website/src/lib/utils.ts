import { InferSelectModel, type courseRegistrationTable } from '@bluedot/db';
import { AxiosError } from 'axios';

export const formatStringToArray = (
  input: string | null | undefined,
  splitBy: string,
): string[] => {
  return input
    ? input
      .split(splitBy)
      .map((o) => o.trim())
      .filter((item) => item !== '')
    : [];
};

export const parseZodValidationError = (err: AxiosError<{ error?: string }>, defaultErrorMessage: string): string => {
  const errorString = err.response?.data?.error;
  if (typeof errorString === 'string' && errorString.startsWith('Invalid request body: ')) {
    try {
      // Extract the JSON array from the error string
      const jsonPart = errorString.replace('Invalid request body: ', '');
      const validationErrors = JSON.parse(jsonPart);
      if (Array.isArray(validationErrors) && validationErrors.length > 0) {
        return validationErrors[0].message || defaultErrorMessage;
      }
      return defaultErrorMessage;
    } catch {
      return defaultErrorMessage;
    }
  }

  return defaultErrorMessage;
};

type CourseRegistration = InferSelectModel<typeof courseRegistrationTable.pg>;
/**
 * Workaround for duplicates existing in rare cases, see issue: https://github.com/bluedotimpact/bluedot/issues/1249
 */
export const stablePickCourseRegistration = (
  courseRegistrations: CourseRegistration[],
): CourseRegistration | undefined => {
  // Sort by ID to ensure stable selection when multiple records exist
  // This ensures we always work with the same registration record
  courseRegistrations.sort((a, b) => a.id.localeCompare(b.id));

  return courseRegistrations[0];
};
