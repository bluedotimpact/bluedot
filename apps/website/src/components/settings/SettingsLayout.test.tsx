import {
  describe,
  expect,
  test,
} from 'vitest';
import { render, screen, within } from '@testing-library/react';
// eslint-disable-next-line import/no-extraneous-dependencies
import '@testing-library/jest-dom';
import SettingsLayout from './SettingsLayout';
import { ROUTES } from '../../lib/routes';

describe('SettingsLayout', () => {
  test('renders complete layout structure with correct active tab', () => {
    const { container } = render(
      <SettingsLayout activeTab="account" route={ROUTES.settingsAccount}>
        <div>Test Content</div>
      </SettingsLayout>,
    );

    // Snapshot only covers Settings side bar menu and main content changing
    expect(container).toMatchSnapshot();
  });

  test('renders different children based on activeTab prop', () => {
    const { rerender } = render(
      <SettingsLayout activeTab="account" route={ROUTES.settingsAccount}>
        <div>Account Settings Content</div>
      </SettingsLayout>,
    );

    // Verify account content renders
    const mainContent = screen.getByRole('main', { name: 'Settings content' });
    expect(within(mainContent).getByText('Account Settings Content')).toBeInTheDocument();

    // Change to courses tab
    rerender(
      <SettingsLayout activeTab="courses" route={ROUTES.settingsCourses}>
        <div>Courses Settings Content</div>
      </SettingsLayout>,
    );

    // Verify courses content renders
    expect(within(mainContent).getByText('Courses Settings Content')).toBeInTheDocument();
    expect(within(mainContent).queryByText('Account Settings Content')).not.toBeInTheDocument();
  });
});
