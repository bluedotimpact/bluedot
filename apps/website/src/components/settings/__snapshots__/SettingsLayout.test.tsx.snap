// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SettingsLayout > renders complete layout structure with correct active tab 1`] = `
<div>
  <div
    class="breadcrumbs bg-color-canvas border-b border-color-divider w-full py-space-between"
  >
    <nav
      aria-label="Breadcrumbs"
      class="breadcrumbs__nav section-base flex flex-row justify-between"
    >
      <ol
        class="breadcrumbs__list flex"
      >
        <li
          class="breadcrumbs__item flex items-center"
        >
          <a
            class="bluedot-a not-prose breadcrumbs__link no-underline"
            href="/"
            tabindex="0"
          >
            Home
          </a>
          <span
            aria-hidden="true"
            class="breadcrumbs__separator mx-2"
          >
            &gt;
          </span>
        </li>
        <li
          class="breadcrumbs__item flex items-center"
        >
          <a
            class="bluedot-a not-prose breadcrumbs__link no-underline"
            href="/settings"
            tabindex="0"
          >
            Settings
          </a>
          <span
            aria-hidden="true"
            class="breadcrumbs__separator mx-2"
          >
            &gt;
          </span>
        </li>
        <li
          class="breadcrumbs__item flex items-center"
        >
          <a
            class="bluedot-a not-prose breadcrumbs__link no-underline opacity-50"
            href="/settings/account"
            tabindex="0"
          >
            Account
          </a>
        </li>
      </ol>
    </nav>
  </div>
  <div
    class="container mx-auto px-4 py-6"
  >
    <div
      class="lg:grid lg:grid-cols-12 lg:gap-8"
    >
      <aside
        aria-label="Settings sidebar"
        class="lg:col-span-3 mb-6 lg:max-w-[320px]"
      >
        <h2
          class="bluedot-h2 not-prose mb-4 font-semibold text-size-lg leading-[1.4]"
        >
          Settings
        </h2>
        <nav
          aria-label="Settings navigation"
        >
          <ul
            class="flex gap-2 flex-wrap lg:block lg:space-y-1"
          >
            <li>
              <a
                aria-label="Account settings"
                class="
                  w-auto lg:w-full h-11 px-4 py-3 rounded-lg transition-colors flex items-center
                  bg-[rgba(0,85,255,0.08)] text-bluedot-darker font-semibold
                "
                href="/settings/account"
                tabindex="0"
              >
                <span
                  class="flex items-center gap-3 w-full"
                >
                  <p
                    class="bluedot-p not-prose text-[13px] leading-[22px] flex-grow !m-0 font-semibold"
                  >
                    Account
                  </p>
                </span>
              </a>
            </li>
            <li>
              <a
                aria-label="Courses settings"
                class="
                  w-auto lg:w-full h-11 px-4 py-3 rounded-lg transition-colors flex items-center
                  text-bluedot-darker hover:bg-[rgba(0,85,255,0.04)]
                "
                href="/settings/courses"
                tabindex="0"
              >
                <span
                  class="flex items-center gap-3 w-full"
                >
                  <p
                    class="bluedot-p not-prose text-[13px] leading-[22px] flex-grow !m-0 font-normal"
                  >
                    Courses
                  </p>
                </span>
              </a>
            </li>
            <li>
              <a
                aria-label="Community settings (New feature)"
                class="
                  w-auto lg:w-full h-11 px-4 py-3 rounded-lg transition-colors flex items-center
                  text-bluedot-darker hover:bg-[rgba(0,85,255,0.04)]
                "
                href="/settings/community"
                tabindex="0"
              >
                <span
                  class="flex items-center gap-3 w-full"
                >
                  <p
                    class="bluedot-p not-prose text-[13px] leading-[22px] flex-grow !m-0 font-normal"
                  >
                    Community
                  </p>
                  <span
                    aria-label="New feature"
                    class="text-[10px] px-1.5 py-0.5 rounded-[5px] uppercase font-bold bg-[rgba(0,85,255,0.08)] text-[#0037FF]"
                  >
                    NEW
                  </span>
                </span>
              </a>
            </li>
          </ul>
        </nav>
      </aside>
      <main
        aria-label="Settings content"
        class="lg:col-span-9"
      >
        <div
          class="container-lined"
          role="region"
        >
          <div>
            Test Content
          </div>
        </div>
      </main>
    </div>
  </div>
</div>
`;
